File: clients/android/app/src/main/java/tv/plurx/app/player/PreparedSwitchMeasurement.kt
package tv.plurx.app.player

/**
 * Measuring a prepared quality switch, with no player in it.
 *
 * Every function here is pure and every value is a plain number the caller has
 * already read off a listener callback. That is deliberate in two directions.
 * It makes the arithmetic — the window edges, the two-sided sum, the counter
 * reset — reachable by the unit lane, which is where this programme's defects
 * have actually lived. And it makes it structurally impossible for the
 * instrument to change what it measures: nothing below touches an `ExoPlayer`,
 * schedules anything, or blocks, so recording a switch cannot alter the timing
 * of one.
 *
 * Nothing here decides anything either. The rows it produces are advisory, in
 * the strict sense Paul's standing rule means: no caller reads them back, and
 * an unmeasurable window is reported as unmeasured rather than as a failure.
 * The bar these readings are compared against lives in
 * `docs/playback-control/QUALITY-SWITCH-CONTINUITY-RESULTS.md`, not in code.
 */

/**
 * Two seconds *either side* of the commit, so a four-second window.
 *
 * A prepared switch spans two pipelines: the predecessor's dropped-frame
 * counter stops at the commit and the successor's starts there. Taking one
 * player's delta across the swap would measure half the window and report it
 * as the whole, which is why [preparedSwitchCounterDelta] takes two series.
 */
internal const val PREPARED_SWITCH_WINDOW_MS = 2_000L

/** How many samples either side are kept. A bound, so an eight-hour film cannot grow it. */
internal const val PREPARED_SWITCH_SAMPLES_MAX = 32

/**
 * One reading of one cumulative counter: the monotonic clock it was taken at
 * and the count at that instant.
 */
internal data class PreparedSwitchSample(val atMs: Long, val count: Long)

/**
 * What a window came to. [count] is null when the window could not be
 * measured at all — never zero, because a zero nobody observed and a zero
 * that was observed are different facts and only one of them clears the bar.
 */
internal data class PreparedSwitchWindow(
    val count: Long?,
    val beforeMs: Long,
    val afterMs: Long,
    val windowMs: Long,
) {
    val coveredMs: Long get() = beforeMs + afterMs
    val spanMs: Long get() = windowMs * 2
}

/**
 * The delta of one cumulative counter over one side of the window.
 *
 * Null with fewer than two readings inside it: a delta needs two, and
 * inventing the missing one is exactly how an instrument comes to certify a
 * window it never saw.
 */
private fun side(
    samples: List<PreparedSwitchSample>,
    fromMs: Long,
    toMs: Long,
): Pair<Long, Long>? {
    val kept = samples.filter { it.atMs in fromMs..toMs }.sortedBy { it.atMs }
    if (kept.size < 2) return null
    val first = kept.first()
    val last = kept.last()
    // A counter that went backwards was reset — a new pipeline, a codec
    // reconfiguration. Everything it has counted since the reset is inside the
    // window, so the reading is the counter itself and never a negative delta.
    val delta = if (last.count >= first.count) last.count - first.count else last.count
    return maxOf(0L, delta) to (last.atMs - first.atMs)
}

/**
 * The two-sided counter delta around a commit.
 *
 * [before] is the predecessor's series and [after] the successor's; each is
 * clipped to its own side of [commitAtMs] so a stale sample cannot be counted
 * twice. Either may be empty, which is a real outcome rather than an error.
 */
internal fun preparedSwitchCounterDelta(
    before: List<PreparedSwitchSample>,
    after: List<PreparedSwitchSample>,
    commitAtMs: Long,
    windowMs: Long = PREPARED_SWITCH_WINDOW_MS,
): PreparedSwitchWindow {
    val bound = if (windowMs > 0) windowMs else PREPARED_SWITCH_WINDOW_MS
    val left = side(before, commitAtMs - bound, commitAtMs)
    val right = side(after, commitAtMs, commitAtMs + bound)
    if (left == null && right == null) {
        return PreparedSwitchWindow(count = null, beforeMs = 0, afterMs = 0, windowMs = bound)
    }
    return PreparedSwitchWindow(
        count = (left?.first ?: 0L) + (right?.first ?: 0L),
        beforeMs = left?.second ?: 0L,
        afterMs = right?.second ?: 0L,
        windowMs = bound,
    )
}

/**
 * How many of [events] fall inside the window around [commitAtMs].
 *
 * An event stream is not a counter and must not be reduced like one. Two
 * readings are needed to take a counter's delta, so a window with one reading
 * is unmeasured — but a window with no *events* in it is a measured zero, as
 * long as the listener was attached across it, which for both pipelines it is
 * by construction. Treating underruns as a counter would report "not measured"
 * for exactly the clean switch the bar is about.
 */
internal fun preparedSwitchEventCount(
    events: List<Long>,
    commitAtMs: Long,
    windowMs: Long = PREPARED_SWITCH_WINDOW_MS,
): Long {
    val bound = if (windowMs > 0) windowMs else PREPARED_SWITCH_WINDOW_MS
    return events.count { it in (commitAtMs - bound)..(commitAtMs + bound) }.toLong()
}

/**
 * Tap to the successor's first frame, in wall time.
 *
 * Null rather than a negative number: two clocks that disagree are not a
 * measurement. Reported, never judged — §1's bar says nothing about it.
 */
internal fun preparedSwitchVisibleInMs(tappedAtMs: Long?, firstFrameAtMs: Long?): Long? {
    if (tappedAtMs == null || firstFrameAtMs == null) return null
    if (firstFrameAtMs < tappedAtMs) return null
    return firstFrameAtMs - tappedAtMs
}

private fun seconds(ms: Long): String = String.format(java.util.Locale.US, "%.1f s", ms / 1_000.0)

/** One wording, shared with the web and Apple ledgers. */
internal fun preparedSwitchFramesRow(window: PreparedSwitchWindow?): String {
    val count = window?.count ?: return "Not measured"
    return "$count dropped · ±${seconds(window.windowMs)} · " +
        "${seconds(window.coveredMs)} of ${seconds(window.spanMs)} sampled"
}

/**
 * The audio row for a platform that observes discrete events. The window is
 * stated but no coverage clause is: a listener attached across the window has
 * seen all of it, and printing "4.0 s of 4.0 s sampled" for an event stream
 * would be describing a counter this is not.
 */
internal fun preparedSwitchAudioUnderrunRow(
    count: Long,
    windowMs: Long = PREPARED_SWITCH_WINDOW_MS,
): String {
    val word = if (count == 1L) "audio underrun" else "audio underruns"
    return "$count $word · ±${seconds(windowMs)}"
}

internal fun preparedSwitchVisibleRow(ms: Long?): String = if (ms == null) "Not measured" else "$ms ms"

/**
 * What a prepared switch measured, as the three strings the ledger shows.
 *
 * A value object, so the player can hand one to the info panel and to the
 * developer screen without either being able to change it.
 */
internal data class PreparedSwitchReading(
    val frames: String,
    val audio: String,
    val visibleIn: String,
) {
    companion object {
        val unmeasured = PreparedSwitchReading(
            frames = "Not measured",
            audio = "Not measured",
            visibleIn = "Not measured",
        )
    }
}

/**
 * The recorder the player pushes into.
 *
 * Deliberately not thread-safe and deliberately dumb: the controller owns it
 * and touches it from the player's own scope, and every method is a push onto
 * a bounded list. The reading is computed on demand, which means a closed info
 * panel costs nothing and a commit costs three list appends.
 */
internal class PreparedSwitchRecorder {
    /**
     * Samples are keyed by the pipeline that reported them rather than by a
     * before/after flag, because a second directed change in the same session
     * makes the first switch's successor the second switch's predecessor. A
     * flag captured at prepare time would then file both pipelines on the same
     * side and add a switch to itself.
     */
    private class Series {
        val frames = ArrayDeque<PreparedSwitchSample>()
        val underruns = ArrayDeque<Long>()
        var droppedTotal = 0L
    }

    private val series = LinkedHashMap<Int, Series>()

    /** Null until a commit has happened, which is what a fresh player looks like. */
    var commitAtMs: Long? = null
        private set
    private var predecessor: Int? = null
    private var successor: Int? = null
    private var tappedAtMs: Long? = null
    private var firstFrameAtMs: Long? = null

    private fun seriesFor(pipeline: Int): Series {
        series[pipeline]?.let { return it }
        // Two pipelines matter and a third means the switch being measured is
        // over. Evicting the oldest keeps this bounded without any timer.
        while (series.size >= 3) series.remove(series.keys.first())
        return Series().also { series[pipeline] = it }
    }

    private fun push(into: ArrayDeque<PreparedSwitchSample>, atMs: Long, count: Long) {
        into.addLast(PreparedSwitchSample(atMs, count))
        while (into.size > PREPARED_SWITCH_SAMPLES_MAX) into.removeFirst()
    }

    /**
     * `AnalyticsListener.onDroppedVideoFrames` reports an INCREMENT since the
     * last call, not a total, so the running sum is kept here.
     */
    fun noteDroppedFrames(atMs: Long, dropped: Int, pipeline: Int) {
        val entry = seriesFor(pipeline)
        entry.droppedTotal += maxOf(0, dropped).toLong()
        push(entry.frames, atMs, entry.droppedTotal)
    }

    /**
     * `AnalyticsListener.onAudioUnderrun` is an event, not a counter. It is
     * recorded as the instant it happened and counted by
     * [preparedSwitchEventCount], so a window with none in it is a measured
     * zero rather than an unmeasured window.
     */
    fun noteAudioUnderrun(atMs: Long, pipeline: Int) {
        val entry = seriesFor(pipeline)
        entry.underruns.addLast(atMs)
        while (entry.underruns.size > PREPARED_SWITCH_SAMPLES_MAX) entry.underruns.removeFirst()
    }

    /** The successor has the surface. Assignments only; nothing is read back. */
    fun noteCommit(atMs: Long, tappedAtMs: Long?, predecessor: Int, successor: Int) {
        commitAtMs = atMs
        this.predecessor = predecessor
        this.successor = successor
        this.tappedAtMs = tappedAtMs
        firstFrameAtMs = null
    }

    /** The successor rendered. Recorded once; a later frame does not move it. */
    fun noteFirstFrame(atMs: Long) {
        if (commitAtMs != null && firstFrameAtMs == null) firstFrameAtMs = atMs
    }

    /** The three rows, computed on demand. */
    fun reading(): PreparedSwitchReading {
        val commit = commitAtMs ?: return PreparedSwitchReading.unmeasured
        val before = predecessor?.let { series[it] }
        val after = successor?.let { series[it] }
        return PreparedSwitchReading(
            frames = preparedSwitchFramesRow(
                preparedSwitchCounterDelta(
                    before?.frames?.toList() ?: emptyList(),
                    after?.frames?.toList() ?: emptyList(),
                    commit,
                ),
            ),
            audio = preparedSwitchAudioUnderrunRow(
                preparedSwitchEventCount(
                    (before?.underruns?.toList() ?: emptyList()).filter { it <= commit } +
                        (after?.underruns?.toList() ?: emptyList()).filter { it >= commit },
                    commit,
                ),
            ),
            visibleIn = preparedSwitchVisibleRow(
                preparedSwitchVisibleInMs(tappedAtMs, firstFrameAtMs),
            ),
        )
    }
}

/**
 * What an audible-seam measurement would need on this platform, and whether
 * each condition currently holds.
 *
 * Advisory in the strict sense: there is no switch attached to any of these,
 * nothing reads them back, and a condition that is not met never stops a
 * quality change, never disables the prepared handoff, and never turns a row
 * into a failure. Android's underrun callback is the one audio observation that
 * costs the audible path nothing, which is why it is the one that ships.
 */
internal fun preparedSwitchAudioRequirements(
    underrunCallbackAvailable: Boolean,
): List<Triple<String, String, Boolean>> = listOf(
    Triple(
        "Audio-sink underrun reporting",
        "Met by Media3's AnalyticsListener. A pure observation of the audio sink: " +
            "it adds no processing to the audible path and cannot mute it.",
        underrunCallbackAvailable,
    ),
    Triple(
        "Waveform gap detection",
        "Not available. Reading the decoded waveform needs an audio processor in the " +
            "sink chain, which is the audible path itself; a processor that stalls is a " +
            "silence the viewer hears. The underrun count above is what ships instead.",
        false,
    ),
)

File: clients/apple/Sources/PreparedSwitchMeasurement.swift
import Foundation

/// Measuring a prepared quality switch, with no AVFoundation in it.
///
/// Every function here is pure and every input is a plain number the caller has
/// already read off an access log. That is deliberate in two directions. It puts
/// the arithmetic — the window edges, the two-sided sum, the counter reset —
/// inside reach of the unit lane, which is where this programme's defects have
/// actually lived. And it makes it structurally impossible for the instrument to
/// be the thing it measures: nothing below touches a player, an item, an output
/// or a clock, so recording a switch cannot change the timing of one.
///
/// Nothing here decides anything either. The rows are advisory in the strict
/// sense Paul's standing rule means: no caller reads them back, no capability
/// consults them, and a window that could not be measured is reported as
/// unmeasured rather than as a failure. The bar these readings are compared
/// against lives in `docs/playback-control/QUALITY-SWITCH-CONTINUITY-RESULTS.md`.
enum PreparedSwitchMeasurement {
    /// Two seconds *either side* of the commit, so a four-second window.
    ///
    /// A prepared switch spans two `AVPlayerItem`s, and each keeps its own
    /// access log: the predecessor's counters stop at `replaceCurrentItem` and
    /// the successor's start there. Reading one item's delta across the swap
    /// would measure half the window and report it as the whole.
    static let windowMs = 2_000

    /// How many readings each side keeps. A bound, so a long film cannot grow it.
    static let samplesMax = 32

    /// One reading of one cumulative counter.
    struct Sample: Equatable {
        let atMs: Int
        let count: Int

        init(atMs: Int, count: Int) {
            self.atMs = atMs
            self.count = count
        }
    }

    /// What a window came to. `count` is nil when the window could not be
    /// measured at all — never zero, because a zero nobody observed and a zero
    /// that was observed are different facts and only one of them clears a bar.
    struct Window: Equatable {
        let count: Int?
        let beforeMs: Int
        let afterMs: Int
        let windowMs: Int

        var coveredMs: Int { beforeMs + afterMs }
        var spanMs: Int { windowMs * 2 }
    }

    /// The delta of one cumulative counter over one side of the window.
    ///
    /// Nil with fewer than two readings inside it: a delta needs two, and
    /// inventing the missing one is how an instrument comes to certify a window
    /// it never saw.
    private static func side(
        _ samples: [Sample],
        from: Int,
        to: Int
    ) -> (delta: Int, coveredMs: Int)? {
        let kept = samples.filter { $0.atMs >= from && $0.atMs <= to }
            .sorted { $0.atMs < $1.atMs }
        guard let first = kept.first, let last = kept.last, kept.count >= 2 else { return nil }
        // A counter that went backwards was reset — a new item, a renderer
        // rebuild. Everything counted since the reset is inside the window, so
        // the reading is the counter itself and never a negative delta.
        let delta = last.count >= first.count ? last.count - first.count : last.count
        return (max(0, delta), last.atMs - first.atMs)
    }

    /// The two-sided counter delta around a commit.
    ///
    /// `before` is the predecessor item's series and `after` the successor's;
    /// each is clipped to its own side of `commitAtMs` so one reading cannot be
    /// counted twice. Either may be empty, which is a real outcome.
    static func counterDelta(
        before: [Sample],
        after: [Sample],
        commitAtMs: Int,
        windowMs: Int = PreparedSwitchMeasurement.windowMs
    ) -> Window {
        let bound = windowMs > 0 ? windowMs : PreparedSwitchMeasurement.windowMs
        let left = side(before, from: commitAtMs - bound, to: commitAtMs)
        let right = side(after, from: commitAtMs, to: commitAtMs + bound)
        if left == nil, right == nil {
            return Window(count: nil, beforeMs: 0, afterMs: 0, windowMs: bound)
        }
        return Window(
            count: (left?.delta ?? 0) + (right?.delta ?? 0),
            beforeMs: left?.coveredMs ?? 0,
            afterMs: right?.coveredMs ?? 0,
            windowMs: bound
        )
    }

    /// Tap to the successor's first frame, in wall time. Nil rather than a
    /// negative number: two clocks that disagree are not a measurement.
    static func visibleInMs(tappedAtUnixMs: Int?, firstFrameUnixMs: Int?) -> Int? {
        guard let tappedAtUnixMs, let firstFrameUnixMs else { return nil }
        guard firstFrameUnixMs >= tappedAtUnixMs else { return nil }
        return firstFrameUnixMs - tappedAtUnixMs
    }

    private static func seconds(_ ms: Int) -> String {
        String(format: "%.1f s", Double(ms) / 1_000)
    }

    /// One wording, shared with the web and Android ledgers.
    static func framesRow(_ window: Window?) -> String {
        guard let window, let count = window.count else { return "Not measured" }
        return "\(count) dropped · ±\(seconds(window.windowMs)) · "
            + "\(seconds(window.coveredMs)) of \(seconds(window.spanMs)) sampled"
    }

    /// The audio row. Apple's observation is the access log's own stall count,
    /// which is a cumulative counter and so reduces the same way the frames do.
    static func audioRow(_ window: Window?) -> String {
        guard let window, let count = window.count else { return "Not measured" }
        let word = count == 1 ? "stall" : "stalls"
        return "\(count) access-log \(word) · ±\(seconds(window.windowMs)) · "
            + "\(seconds(window.coveredMs)) of \(seconds(window.spanMs)) sampled"
    }

    static func visibleRow(_ ms: Int?) -> String {
        guard let ms else { return "Not measured" }
        return "\(ms) ms"
    }

    /// The three rows the ledger shows.
    struct Reading: Equatable {
        let frames: String
        let audio: String
        let visibleIn: String

        static let unmeasured = Reading(
            frames: "Not measured",
            audio: "Not measured",
            visibleIn: "Not measured"
        )
    }

    /// The recorder the controller pushes into.
    ///
    /// Deliberately dumb: two bounded arrays and four assignments. The reading
    /// is computed on demand, so a closed info panel costs nothing and a commit
    /// costs three stores.
    struct Recorder {
        private var before: [Sample] = []
        private var after: [Sample] = []
        private var beforeStalls: [Sample] = []
        private var afterStalls: [Sample] = []
        private(set) var commitAtMs: Int?
        private var tappedAtUnixMs: Int?
        private var firstFrameUnixMs: Int?

        private static func push(_ into: inout [Sample], _ sample: Sample) {
            // A repeated reading of an unchanged counter at the same instant is
            // noise; a repeated instant would also make `coveredMs` lie.
            if let last = into.last, last.atMs == sample.atMs { into.removeLast() }
            into.append(sample)
            if into.count > samplesMax { into.removeFirst(into.count - samplesMax) }
        }

        /// A reading of the item that is currently on the layer. Which side of
        /// the commit it lands on is decided by whether a commit has happened
        /// yet, because `replaceCurrentItem` is the instant the item changes.
        mutating func note(atMs: Int, droppedFrames: Int?, stalls: Int?) {
            let committed = commitAtMs != nil
            if let droppedFrames {
                if committed { Self.push(&after, Sample(atMs: atMs, count: droppedFrames)) }
                else { Self.push(&before, Sample(atMs: atMs, count: droppedFrames)) }
            }
            if let stalls {
                if committed { Self.push(&afterStalls, Sample(atMs: atMs, count: stalls)) }
                else { Self.push(&beforeStalls, Sample(atMs: atMs, count: stalls)) }
            }
        }

        /// The viewer tapped. Recorded in wall time because the first frame is.
        mutating func note(tappedAtUnixMs: Int) {
            self.tappedAtUnixMs = tappedAtUnixMs
        }

        /// The successor is on the layer. Assignments only.
        mutating func note(commitAtMs: Int) {
            self.commitAtMs = commitAtMs
            firstFrameUnixMs = nil
        }

        /// The successor rendered. Recorded once; a later frame does not move it.
        mutating func note(firstFrameUnixMs: Int) {
            guard commitAtMs != nil, self.firstFrameUnixMs == nil else { return }
            self.firstFrameUnixMs = firstFrameUnixMs
        }

        /// A new staging. The samples taken for the last switch are not this
        /// one's, and the tap is: the wait for an offer happens before this.
        mutating func reopened() {
            before = after
            after = []
            beforeStalls = afterStalls
            afterStalls = []
            commitAtMs = nil
            firstFrameUnixMs = nil
        }

        func reading() -> Reading {
            guard let commitAtMs else { return .unmeasured }
            return Reading(
                frames: framesRow(counterDelta(before: before, after: after, commitAtMs: commitAtMs)),
                audio: audioRow(
                    counterDelta(before: beforeStalls, after: afterStalls, commitAtMs: commitAtMs)
                ),
                visibleIn: visibleRow(
                    visibleInMs(tappedAtUnixMs: tappedAtUnixMs, firstFrameUnixMs: firstFrameUnixMs)
                )
            )
        }
    }

    /// What an audible-seam measurement would need on this platform, and whether
    /// each condition holds.
    ///
    /// Advisory in the strict sense: there is no switch attached to any of
    /// these, nothing reads them back, and a condition that is not met never
    /// refuses a quality change, never disables the prepared handoff, and never
    /// turns a row into a failure. `AVPlayerItemAccessLog`'s stall count is the
    /// one audio observation that costs the audible path nothing, which is why
    /// it is the one that ships.
    struct Requirement: Identifiable, Equatable {
        let title: String
        let detail: String
        let met: Bool

        var id: String { title }
    }

    static func audioRequirements(accessLogAvailable: Bool) -> [Requirement] {
        [
            Requirement(
                title: "Access-log stall count",
                detail: "Met when the item keeps an access log. A read of a log AVFoundation "
                    + "already writes: it adds nothing to the audio path and cannot mute it.",
                met: accessLogAvailable
            ),
            Requirement(
                title: "Waveform gap detection",
                detail: "Not available. An MTAudioProcessingTap is installed by setting the item's "
                    + "audioMix, which puts a real-time callback of ours inside the audible "
                    + "path on every session — a callback that overruns is a silence the "
                    + "viewer hears. The stall count above is what ships instead.",
                met: false
            ),
        ]
    }
}
